File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode ref = head;
        int size = 0;
        
        while (ref != null) {
            size++;
            ref = ref.next;
        }

        if (size == 1)
            return null;
        
        if (size == n) {
            head = head.next;            
            return head;
        }
        
        ListNode ref1 = head;
        for (int nth = 0; nth < size - n - 1; nth++)
            ref1 = ref1.next;
        
        ref1.next = ref1.next.next;
        
        return head;
    }
}

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        length = 0
        ref = head
        while ref:
            ref = ref.next
            length += 1
        
        if length == 1:
            return None
        
        if length == n:
            return head.next
        
        ref1 = ListNode(-1, head)
        for i in range(length - n - 1):
            head = head.next
        
        head.next = head.next.next
        
        return ref1.next